### PR TITLE
Handle 'Pushback'ed characters correctly when reading stream position

### PR DESCRIPTION
**Pull request by ChrisJefferson**
@laurentbartholdi Found a bug (in #1878) in the parser. While it appeared when multi-argument lambdas were introduced, the deeper problem was correctly handling when characters are pushed back into the stream.

### CODE REVIEW

File: src/scanner.c [hunk 1]
```diff
@@ -1275,10 +1290,27 @@ static inline void GET_CHAR( void ) {
...
-  STATE(In) = &Pushback;
+    assert(!CHAR_PUSHBACK_EMPTY());
+    Pushback = c;
+    RealIn = STATE(In);
+    STATE(In) = &Pushback;
+}
+
+// Get current line position. In the case where we pushed back the last
+// character on the previous line we return the first character of the
+// current line, as cannot retrieve the previous line.
```
Comment (fingolfin) [inline src/scanner.c]: Missing "we" after "as"?

Comment (ChrisJefferson) [reply src/scanner.c]: Fixed (along with some other minor issues which were causing tests to fail)

File: src/scanner.c [hunk 2]
```diff
@@ -135,9 +139,11 @@ void            SyntaxError (
         Pr( "%s", (Int)STATE(Input)->line, 0L );
 
         /* print a '^' pointing to the current position                        */
-        for ( i = 0; i < STATE(In) - STATE(Input)->line - 1; i++ ) {
-          if ( STATE(Input)->line[i] == '\t' )  Pr("\t",0L,0L);
-          else  Pr(" ",0L,0L);
+        Int pos = GetLinePosition();
```
Comment (fingolfin) [inline src/scanner.c]: I like the new `GetLinePosition`, even without the fix, because it helps to make the code more understandable. Nice :-)

File: src/scanner.c [hunk 3]
```diff
@@ -135,9 +139,11 @@ void            SyntaxError (
...
+        for (i = 0; i < pos; i++) {
+            if (STATE(Input)->line[i] == '\t')
+                Pr("\t", 0L, 0L);
+            else  Pr(" ",0L,0L);
```
Comment (fingolfin) [inline src/scanner.c]: Did you run `git clang-format` on this? Just wondering, because `Pr` is in the same line as `else`.

Comment (fingolfin) [reply src/scanner.c]: (since the rest of the function is clearly not clang formatted, it's also fine to leave it like this. I was simply wondering).

File: src/scanner.c [hunk 4]
```diff
@@ -103,6 +103,10 @@ Obj             EndLineHook = 0;
 #endif
 
 
+// Forward declarations
+static Int        GetLinePosition(void);
+static inline Int CHAR_PUSHBACK_EMPTY(void);
```
Comment (fingolfin) [inline src/scanner.c]: Why not move the function definitions here, instead of a forward declaration?

File: src/scanner.c [hunk 5]
```diff
@@ -386,6 +394,7 @@ UInt OpenInput (
 
     /* remember the current position in the current file                   */
     if ( STACK_SIZE(Input) > 0 ) {
+        GAP_ASSERT(CHAR_PUSHBACK_EMPTY());
```
Comment (fingolfin) [inline src/scanner.c]: I was briefly confused by this line, then I realized `CHAR_PUSHBACK_EMPTY()` is tester function returning a boolean. So perhaps call it `IS_CHAR_PUSHBACK_EMPTY` to make it crystal clear?

File: src/scanner.c [hunk 6]
```diff
@@ -1265,6 +1275,11 @@ Char GetLine ( void )
 static Char Pushback = '\0';
 static Char *RealIn;
```
Comment (fingolfin) [inline src/scanner.c]: I realize this is not part of this PR, but: isn't code using these two globals thread unsafe? I.e. if two threads in HPC-GAP were trying to parse a file, they could clash with each other if they both try to "unget" a char simultaneously.

So we should move these into the `GAPState` struct, shouldn't we?

Comment (ChrisJefferson) [reply src/scanner.c]: That is an *extremely* good comment, and I'm slightly embarassed I didn't think of it. I think I've even seen a couple of crashes about this I could never reproduce (parsing weirdly claiming something couldn't be parsed with weird errors)

**Comment by fingolfin** [comment]
Second commit message starts with "Move In and Pushback" -> that should be "RealIn", not "In"

File: src/scanner.c [hunk 7]
```diff
@@ -105,6 +105,66 @@ Obj             EndLineHook = 0;
 
 /****************************************************************************
 **
+*F  GET_CHAR()  . . . . . . . . . . . . . . . . get the next character, local
+**
+**  'GET_CHAR' returns the next character from  the current input file.  This
+**  character is afterwords also available as '*In'.
```
Comment (fingolfin) [inline src/scanner.c]: Hehe, "afterwords" ....  o-> a

File: src/scanner.c [hunk 8]
```diff
@@ -105,6 +105,66 @@ Obj             EndLineHook = 0;
...
+**
+**  For efficiency  reasons 'GET_CHAR' is a  macro that  just  increments the
```
Comment (fingolfin) [inline src/scanner.c]: `GET_CHAR` isn't a macro anymore. I'd just loose the first part of this paragraph and start maybe like this:
>  'GET_CHAR' increments the pointer ...

**Comment by ChrisJefferson** [comment]
Fixed comments. Actually removed comment about being a macro and what the function did, as you can just read it. Added a mention in the comment about pushback, as that was the source of this whole bug.
